File: typescript/src/features/goose/mcp_registry.ts
// Re-implementation of Goose's Model Context Protocol (MCP) Registry

export interface MCPTool {
  name: string;
  description: string;
  inputSchema: any;
  execute: (input: any) => Promise<string>;
}

export class MCPRegistry {
  private tools: Map<string, MCPTool> = new Map();

  public registerTool(tool: MCPTool): void {
    this.tools.set(tool.name, tool);
  }

  public getTool(name: string): MCPTool | undefined {
    return this.tools.get(name);
  }

  public getAllTools(): MCPTool[] {
    return Array.from(this.tools.values());
  }

  public async executeTool(name: string, input: any): Promise<string> {
    const tool = this.tools.get(name);
    if (!tool) {
      throw new Error(`Tool ${name} not found in MCP registry.`);
    }
    return await tool.execute(input);
  }
}

File: go/internal/goose/mcp_registry.go
package goose

// Re-implementation of Goose's Model Context Protocol (MCP) Registry

import (
	"errors"
	"fmt"
)

type MCPTool struct {
	Name        string
	Description string
	InputSchema string
	Execute     func(input string) (string, error)
}

type MCPRegistry struct {
	tools map[string]MCPTool
}

func NewMCPRegistry() *MCPRegistry {
	return &MCPRegistry{
		tools: make(map[string]MCPTool),
	}
}

func (r *MCPRegistry) RegisterTool(tool MCPTool) {
	r.tools[tool.Name] = tool
}

func (r *MCPRegistry) GetTool(name string) (MCPTool, bool) {
	tool, exists := r.tools[name]
	return tool, exists
}

func (r *MCPRegistry) ExecuteTool(name string, input string) (string, error) {
	tool, exists := r.tools[name]
	if !exists {
		return "", errors.New(fmt.Sprintf("Tool %s not found in MCP registry.", name))
	}
	return tool.Execute(input)
}
